beacon_node/lighthouse_network/src/peer_manager/network_behaviour.rs

```diff
@@ -169,26 +170,65 @@ impl<TSpec: EthSpec> NetworkBehaviour for PeerManager<TSpec> {
         }
     }
 
+    fn handle_pending_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _local_addr: &libp2p::Multiaddr,
+        remote_addr: &libp2p::Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        let ip = match remote_addr.iter().next() {
+            Some(libp2p::multiaddr::Protocol::Ip6(ip)) => IpAddr::V6(ip),
+            Some(libp2p::multiaddr::Protocol::Ip4(ip)) => IpAddr::V4(ip),
+            _ => {
+                return Err(ConnectionDenied::new(format!(
+                    "Connection to peer rejected: invalid multiaddr: {remote_addr}"
+                )))
+            }
+        };
+
+        if self.network_globals.peers.read().is_ip_banned(&ip) {
+            return Err(ConnectionDenied::new(format!(
+                "Connection to peer rejected: peer {ip} is banned"
+            )));
+        }
+
+        Ok(())
+    }
+
     fn handle_established_inbound_connection(
         &mut self,
         _connection_id: ConnectionId,
-        _peer: PeerId,
+        peer_id: PeerId,
         _local_addr: &libp2p::Multiaddr,
-        _remote_addr: &libp2p::Multiaddr,
-    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
-        // TODO: we might want to check if we accept this peer or not in the future.
+        remote_addr: &libp2p::Multiaddr,
+    ) -> Result<libp2p::swarm::THandler<Self>, ConnectionDenied> {
+        trace!(self.log, "Inbound connection"; "peer_id" => %peer_id, "multiaddr" => %remote_addr);
+        // We already checked if the peer was banned on `handle_pending_inbound_connection`.
+        if let Some(BanResult::BadScore) = self.ban_status(&peer_id) {
+            return Err(ConnectionDenied::new(
+                "Connection to peer rejected: peer has a bad score",
+            ));
+        }
         Ok(ConnectionHandler)
     }
 
     fn handle_established_outbound_connection(
         &mut self,
         _connection_id: ConnectionId,
-        _peer: PeerId,
-        _addr: &libp2p::Multiaddr,
+        peer_id: PeerId,
+        addr: &libp2p::Multiaddr,
         _role_override: libp2p::core::Endpoint,
     ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
-        // TODO: we might want to check if we accept this peer or not in the future.
-        Ok(ConnectionHandler)
+        trace!(self.log, "Outbound connection"; "peer_id" => %peer_id, "multiaddr" => %addr);
+        match self.ban_status(&peer_id) {
+            Some(cause) => {
+                error!(self.log, "Dialing a banned peer. Re-banning"; "peer_id" => %peer_id);
+                // Re-ban the peer to prevent repeated errors.
+                self.events.push(PeerManagerEvent::Banned(peer_id, vec![]));
+                Err(ConnectionDenied::new(cause))
```

Comment:
somewhat unrelated to this Pr, but I wonder if we are counting this peer in the banned peers per ip :thinking: I kinda suspect we are not

Reply:
@AgeManning maybe you can check @jxs check this: whether we do and if not, whether we care to do. I assume yes in general

Reply:
@jxs and I have had a discussion about this exact code path. My stance is that is could almost be an `unreachable!()`. This should never really happen unless there is a bug in our code or our dependencies. 

We should not dial banned peers and we should not establish connections with them. We should prevent this at an earlier stage. 

Of course there could be some kind of race-condition where we attempt a dial and maybe we process some delayed message that makes the peer banned, but in either case, I dont think we should worry too much about complicated logic in this particular code path. 

That being said, in the earlier stages, like us stopping connections from IPs I agree we should be careful to increment counters for Ips as well as peer-ids when we can. 

I think in this PR, libp2p early on in the inbound connection attempt, gives us an IP but does not negotiate a key-exchange so we don't have access to a peer-id. We prevent these connections based on banned-ips and we don't know which potential PeerId it was that was trying to connect. This is the desired approach I think because we dont want to waste time doing key-exchange with banned ips.

Reply:
Thanks for raising it Diva, and thanks Age for the explanation, updated the code to also add the ip to the banned peer ip's, see https://github.com/sigp/lighthouse/pull/4569/commits/ff888bbecc197f5b4d518f2390cfc390602a77db

Reply:
Let me see if I got this right:
If we land here, either the peer is banned or the ip is banned
- the peer is banned: this is an outbound connection, so this node dialed. Right now dialing means exclusively the peer manager dialing which we "don't do", thus we don't land here.
- the ip is banned: again, we dialed. We either provided a banned ip (we don't) or it was banned while dialing, or another behaviour provided this ip (don't remember our configurations here, but they aren't too relevant). 
 
In any case we don't care because in both we know what is banned

This makes sense to me. My question is still, do we, in the second case, want to mark the peer as banned? (the peer is not banned, just connected via a banned ip) Note that we are emitting an event to ban the peer _outside_ the network behaviour, so baiscally discovery. In this case, this ban will never be lifted - because the peer is not banned to be unbanned

Reply:
Ok so I just checked, `BannedIp` does not ban the peer. This just checks that their known addresses are not banned. My conclussion is that if we do land here for whatever reason this will introduce disparities between what's banned in discovery and what's banned in the peer manager. If you think we still chose to not address this, then at a minimum this comment is in fact not correct:

`// Re-ban the peer to prevent repeated errors.`

There is no re-banning (at the peer manager) occurring here

and should be removed.

About adding the ips to the emitted event, not sure we should do that since we don't know if the ip the peer used to connect is one of the previous ones or not, so we don't know if the current ip is banned.

sp TLDR is:
- emitting this event without banning the peer will introduce disparities between discovery and the peer manager because while we reject the connection, there is no re-banning happening
- do not send the received ip address since we are not sure this ip in particular is banned.
- options:
  - accept this as a fact and adjust wording. There is no rebanning occurring here
  - do not emit the event since this now goes only to discovery + adjust wording. No rebanning here plus no disparities with discovery
  - ban the peer here since it will be banned in discovery and would need to be banned here to be unbanned later. Not a fan of this because it kinda messes around with what banned ip means?
 
What do you guys think? :)

Reply:
Ok yep. I misunderstood this the first time. 

I thought we re-banned the peer in a sub call. 

@jxs also suggested that we remove this event, so I'm in favour of just removing the event. We can log the error and if it ever occurs deal with it then.